src/libDirectoryService/DirectoryService.cpp

```diff
@@ -1127,6 +958,121 @@ bool DirectoryService::ProcessAllPoWConnResponse(
         m_hasAllPoWconns = true;
     }
     cv_allPowConns.notify_all();
+
+    return true;
+}
+
+// Current this is only used by ds. But ideally, 20 ds nodes should
+bool DirectoryService::ProcessAllPoW2Request(
+    const vector<unsigned char>& message, unsigned int offset, const Peer& from)
+{
+    LOG_MARKER();
+    LOG_EPOCH(INFO, to_string(m_mediator.m_currentEpochNum).c_str(),
+              "I am sending AllPow2 to requester");
+
+    lock_guard<mutex> g(m_mutexAllPOW1);
+    lock_guard<mutex> g2(m_mutexAllPOW2);
+    lock_guard<mutex> g3(m_mutexAllPoWConns);
+
+    if (!m_allPoW1s.empty())
+    {
+        m_allPoW2s.clear();
+
+        for (auto i : m_allPoW1s)
+        {
+            //Winner will become DS (leader), thus we should not put in POW2
+            if (m_allPoWConns[i.first] == from)
+            {
+                continue;
+            }
+
+            m_allPoW2s.insert(i);
+        }
+
+        //Add previous DS commitee (oldest one), because it back to normal node and should be collected
+        lock_guard<mutex> g4(m_mediator.m_mutexDSCommitteePubKeys);
+        lock_guard<mutex> g5(m_mediator.m_mutexDSCommitteeNetworkInfo);
+        m_allPoW2s.insert(make_pair(m_mediator.m_DSCommitteePubKeys.back(),
+                                    (boost::multiprecision::uint256_t)1));
+        m_allPoWConns.insert(
+            make_pair(m_mediator.m_DSCommitteePubKeys.back(),
+                      m_mediator.m_DSCommitteeNetworkInfo.back()));
+    }
+    else
+    {
+        //Winner will become DS (leader), thus we should not put in POW2
+        for (auto i : m_allPoW2s)
```

Comment:
Do not use auto-for when you need to modify the container `m_allPoW2s` during the iteration

Reply:
ok